feat: Add a built-in AWS_IAM authorizer for HTTP APIs

harrisonhjones opened the pull request:
Issue #, if available: #1731

*Description of changes:*

This is another approach for adding AWS_IAM support to HTTP APIs created by SAM. Other approaches can be found in #1876 and #1878. This approach is based on a conversation with @jfuss.

In this version a new property is added to the Serverless HTTP API object called `EnableIamAuthorizer`. When set to true an `AWS_IAM` authorizer is automatically added to the HTTP API's (internal) list of authorizers under the key "AWS_IAM". This authorizer is setup to us IAM auth. If enabled (It is not enabled by default) it can be set as the HTTP API's default authorizer and set as the authorizer for any Serverless functions that use the API.

*Description of how you validated changes:*

I created the following test template:

```yaml
AWSTemplateFormatVersion: "2010-09-09"
Description: AWS SAM template with a simple API definition
Transform: AWS::Serverless-2016-10-31
Globals:
  HttpApi:
    Auth:
      EnableIamAuthorizer: true
Resources:
  #######
  # Serverless functions that use the automatically-created AWS::Serverless::HttpApi called "ServerlessHttpApi".
  #######
  # Should have no auth set because there is no default authorizer defined on the API.
  HttpApiFunctionDefaultApiDefaultAuth:
    Type: AWS::Serverless::Function
    Properties:
      Events:
        ApiEvent:
          Type: HttpApi
          Properties:
            Path: /default-api/default-auth
            Method: GET
      Handler: index.handler
      InlineCode:
        "def handler(event, context):\n    return {'body':
        'HttpApiFunctionDefaultApiDefaultAuth', 'statusCode': 200}\n"
      Runtime: python3.8
  # Should have IAM auth set because the IAM authorizer is enabled (see Globals) and set.
  HttpApiFunctionDefaultApiIAMAuth:
    Type: AWS::Serverless::Function
    Properties:
      Events:
        ApiEvent:
          Type: HttpApi
          Properties:
            Path: /default-api/iam-auth
            Method: GET
            Auth:
              Authorizer: AWS_IAM
      Handler: index.handler
      InlineCode:
        "def handler(event, context):\n    return {'body':
        'HttpApiFunctionDefaultApiIAMAuth', 'statusCode': 200}\n"
      Runtime: python3.8
  #######
  # Serverless functions that use a manually-created AWS::Serverless::HttpApi with the IAM authorizer enable and the default auth set to the IAM authorizer.
  #######
  # Should have IAM auth set because the default authorizer is the IAM authorizer.
  HttpApiFunctionCustomApiWithDefaultIamAuthDefaultAuth:
    Type: AWS::Serverless::Function
    Properties:
      Events:
        ApiEvent:
          Type: HttpApi
          Properties:
            Path: /custom-api-with-default-iam-auth/default-auth
            Method: GET
            ApiId: !Ref CustomServerlessHttpApiWithDefaultIamAuth
      Handler: index.handler
      InlineCode:
        "def handler(event, context):\n    return {'body':
        'HttpApiFunctionCustomApiWithDefaultIamAuthDefaultAuth', 'statusCode':
        200}\n"
      Runtime: python3.8
  # Should have no auth (override the default authorizer).
  HttpApiFunctionCustomApiWithDefaultIamAuthNoneAuth:
    Type: AWS::Serverless::Function
    Properties:
      Events:
        ApiEvent:
          Type: HttpApi
          Properties:
            Path: /custom-api-with-default-iam-auth/none-auth
            Method: GET
            ApiId: !Ref CustomServerlessHttpApiWithDefaultIamAuth
            Auth:
              Authorizer: NONE
      Handler: index.handler
      InlineCode:
        "def handler(event, context):\n    return {'body':
        'HttpApiFunctionCustomApiWithDefaultIamAuthNoneAuth', 'statusCode':
        200}\n"
      Runtime: python3.8
  # Should have IAM auth set because the authorizer is set to the IAM authorizer.
  HttpApiFunctionCustomApiWithDefaultIamAuthIamAuth:
    Type: AWS::Serverless::Function
    Properties:
      Events:
        ApiEvent:
          Type: HttpApi
          Properties:
            Path: /custom-api-with-default-iam-auth/iam-auth
            Method: GET
            ApiId: !Ref CustomServerlessHttpApiWithDefaultIamAuth
            Auth:
              Authorizer: AWS_IAM
      Handler: index.handler
      InlineCode:
        "def handler(event, context):\n    return {'body':
        'HttpApiFunctionCustomApiWithDefaultIamAuthIamAuth', 'statusCode':
        200}\n"
      Runtime: python3.8
  # Serverless Http Api with an IAM authorizer (default) and a None authorizer.
  CustomServerlessHttpApiWithDefaultIamAuth:
    Type: AWS::Serverless::HttpApi
    Properties:
      FailOnWarnings: true
      Auth:
        EnableIamAuthorizer: true
        DefaultAuthorizer: AWS_IAM

  #######
  # Serverless functions that use a manually-created AWS::Serverless::HttpApi with the IAm authorizer enabled but the default auth unset.
  #######
  # Should have no auth set because the default authorizer is not set.
  HttpApiFunctionCustomApiWithNoDefaultAuthDefaultAuth:
    Type: AWS::Serverless::Function
    Properties:
      Events:
        ApiEvent:
          Type: HttpApi
          Properties:
            Path: /custom-api-with-no-default-auth/default-auth
            Method: GET
            ApiId: !Ref CustomServerlessHttpApiWithNoDefault
      Handler: index.handler
      InlineCode:
        "def handler(event, context):\n    return {'body':
        'HttpApiFunctionCustomApiWithNoDefaultAuthDefaultAuth', 'statusCode':
        200}\n"
      Runtime: python3.8
  # Should have IAM auth set because the authorizer is set to the IAM authorizer.
  HttpApiFunctionCustomApiWithNoDefaultIamAuth:
    Type: AWS::Serverless::Function
    Properties:
      Events:
        ApiEvent:
          Type: HttpApi
          Properties:
            Path: /custom-api-with-no-default-auth/iam-auth
            Method: GET
            ApiId: !Ref CustomServerlessHttpApiWithNoDefault
            Auth:
              Authorizer: AWS_IAM
      Handler: index.handler
      InlineCode:
        "def handler(event, context):\n    return {'body':
        'HttpApiFunctionCustomApiWithNoDefaultIamAuth', 'statusCode': 200}\n"
      Runtime: python3.8
  # Serverless Http Api with the IAM authorizer enabled but no default authorizer set.
  CustomServerlessHttpApiWithNoDefault:
    Type: AWS::Serverless::HttpApi
    Properties:
      FailOnWarnings: true
      Auth:
        EnableIamAuthorizer: true
```

I then compiled it using:

python ./bin/sam-translate.py --template-file=../project/template.yml
I then manually deployed it using CloudFormation in my personal AWS account.

Once deployed I validated all the routes had the expected auth type:

*Default API*

1. https://ptwccx1qv4.execute-api.us-east-1.amazonaws.com/default-api/default-auth - has no auth
1. https://ptwccx1qv4.execute-api.us-east-1.amazonaws.com/default-api/iam-auth - has IAM auth

*Custom API with no default auth*

1. https://warmewfp00.execute-api.us-east-1.amazonaws.com/custom-api-with-no-default-auth/default-auth - has no auth
1. https://warmewfp00.execute-api.us-east-1.amazonaws.com/custom-api-with-no-default-auth/iam-auth - has IAM auth

*Custom API with default IAM auth*

1. https://fj5vzsu863.execute-api.us-east-1.amazonaws.com/custom-api-with-default-iam-auth/default-auth - has IAM auth
1. https://fj5vzsu863.execute-api.us-east-1.amazonaws.com/custom-api-with-default-iam-auth/iam-auth - has IAM auth
1. https://fj5vzsu863.execute-api.us-east-1.amazonaws.com/custom-api-with-default-iam-auth/none-auth - has no auth

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [ ] Update documentation
- [x] Verify transformed template deploys and application functions as expected

*Examples?*

Please reach out in the comments, if you want to add an example. Examples will be 
added to `sam init` through https://github.com/awslabs/aws-sam-cli-app-templates/

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jfuss commented:
@harrisonhjones I think this is probably the best option from the 3 you have proposed. I didn't do a deep review of the code, but I think the setup you have is good. I am having another member of my team (@hawflau) look at this as well, just so we have some agreement on the approach.

hawflau commented:
@harrisonhjones This option looks good to me.
I compared with the other two options you provided, but it seems they require altering things that customers might currently rely on. We should avoid this.
The code implementation in this option also looks fine. We just probably need more tests. And when we roll this out, we need docs and examples to let customers know how to use IAM auth

harrisonhjones commented:
> @harrisonhjones This option looks good to me.

:+1:

> I compared with the other two options you provided, but it seems they require altering things that customers might currently rely on. We should avoid this.

I agree.

> The code implementation in this option also looks fine. We just probably need more tests.

Got it. Will add tests.

> And when we roll this out, we need docs and examples to let customers know how to use IAM auth

Where do I add that? In this package or elsewhere? I see an "Update Documentation" checkbox but after a quick scan of the repo I don't see how documentation is supposed to be updated.

harrisonhjones commented:
# 2/18/2021 Update:

- ➕ Added tests
- ✅ `make pr` passes
- ✅ integration test: `pytest --no-cov integration/single/test_function_with_http_api_and_auth.py` passes
- ✅existing integration test: `pytest --no-cov integration/single/test_basic_http_api.py` passes
- ⚠️ confirmed I don't like writing in Python.

jfuss commented on samtranslator/model/api/http_api_generator.py:
```diff
@@ -489,14 +489,21 @@ def _set_default_authorizer(self, open_api_editor, authorizers, default_authoriz
             )
 
-    def _get_authorizers(self, authorizers_config, default_authorizer=None):
+    def _get_authorizers(self, authorizers_config, enable_iam_authorizer=False):
```
Just making sure I understand here. It seems like `default_authorizer` was never used in this function so it should be safe to updated this. I just want to double check what I am seeing/understanding to make sure there is no accidentally side affect here.

harrisonhjones replied:
yea. I was a bit surprised when I noticed it appeared to be unused. As far as I can tell it was unused.

mgrandis replied:
Used in `ApiGenerator` so maybe it was copied from there.

jfuss commented on samtranslator/model/api/http_api_generator.py:
```diff
@@ -489,14 +489,21 @@ def _set_default_authorizer(self, open_api_editor, authorizers, default_authoriz
         """
         authorizers = {}
 
+        if enable_iam_authorizer:
```
I haven't got to the test yet but we should make sure we testing that handles different arguments. Python objects are all truthy, so what I want to make sure we validate is that the end overall experience handles things like "empty resource value, true is turned into a bool from the template, same with false, what happens when a customer sends a string instead of a bool type. 

My main concern is how this might behave. SAM runs before CFN will do anything to a template, so we get a lot of 'raw' data and can lead to small bugs that are harder to unwind/support.

mgrandis replied:
Should maybe switch to `enable_iam_authorizer == True`

harrisonhjones replied:
Why that over the implied `== True`? (Genuinely asking; Python is not my strong suit)

mgrandis replied:
We could pass `"False"` in this value for example and it would pass the if condition.
```
>>> a = "False"
>>> "foo" if a else "bar"
'foo'
```
We should actually use `is True` instead of `== True`:
```
>>> 1 == True
True
```

harrisonhjones replied:
Got it. Fairly easy change.

jfuss commented on samtranslator/model/eventsources/push.py:
```diff
@@ -1167,36 +1167,51 @@ def _add_auth_to_openapi_integration(self, api, editor):
-                ),
-            )
+        # The IAM authorizer is built-in and not defined as a regular Authorizer.
+        enable_iam_authorizer = api_auth and api_auth.get("EnableIamAuthorizer")
```
If `EnabledIamAuthorizer` is `yes` or `"true"` or `FN:if...` this expression would trigger to be true and cause us to enable the `IAM_AUTH`. Somewhere we should have validation on the value of `EnbableIamAuthorizer`. I think @mgrandis is looking/thinking about this area already. I would like to make sure we consider any validation we may need to do on this pr before merging. I don't have the best understanding of what should be done in this area. 

@mgrandis Could you take a look through this and suggest how we might add this so we keep things consistent with what you are working through now?

mgrandis replied:
We could maybe go with:
`enable_iam_authorizer = api_auth and api_auth.get("EnableIamAuthorizer", False) is True`

If EnableIamAuthorizer is missing, default to False and check that the value is indeed a Boolean True

harrisonhjones replied:
Works for me

jfuss commented on tests/model/api/test_http_api_generator.py:
```diff
@@ -76,6 +77,57 @@ def test_auth_intrinsic_default_auth(self):
         with pytest.raises(InvalidResourceException):
             HttpApiGenerator(**self.kwargs)._construct_http_api()
 
+    def test_auth_iam_enabled(self):
```
This goes back to my validation comment. This might not be the right place to assert that happens but we should have some negative test cases (string passed, dict, etc). The main reason is that CFN send json data to a CFN Macro, which is how SAM runs, and CFN doesn't do any validation on that input. So we need to make sure we are guarding our code properly from input way may receive in that json data.

mgrandis replied:
Agree, on a general note we need to test against many more bad/wrong inputs

harrisonhjones replied:
Do y'all want me to write more tests? If so, I'd appreciate some guidance on which tests to write.

mgrandis replied:
You could add a `test_auth_iam_not_enabled_with_unsupported_values(self)`
It would loop on an array of values and check for each one that no IAM auth is added.
Examples of non boolean values that we could get as input:
`unsupported_values = [1, "", [], {}, { "Ref": "MyVar" }, "True", None]`

harrisonhjones replied:
Got it. Fairly easy change.

harrisonhjones replied:
Turns out this caught an issue so great suggestion!

jfuss commented:
@hawflau or @mgrandis Could one of take a look through this as well?

jfuss commented:
@harrisonhjones 

> Where do I add that? In this package or elsewhere? I see an "Update Documentation" checkbox but after a quick scan of the repo I don't see how documentation is supposed to be updated.

I would need to double check, but I believe we generate docs based on https://github.com/aws/serverless-application-model/blob/develop/versions/2016-10-31.md. I will handle the communication with our doc writer about getting these things added, but updating that file will help in our process.

mgrandis commented:
Sorry about the late review @harrisonhjones!

I have mixed feelings about adding a new property `EnableIamAuthorizer` to `Auth` rather than a new value for `Authorizers`.

Couldn't we add a new `IamAuthorizer` with that `EnableIamAuthorizer` property in it instead?
It would be future proof if we were to add a new option/property to configure the `IamAuthorizer` in the future.

jfuss commented:
@mgrandis The main reason @harrisonhjones and I landed on this design was because the IAM Auth is a static configuration. Forcing a customer to specify `Authorizers` seemed like unnecessary set. `Authorizers` is also a set a different objects, which can be confusing. So we elected to push this up to an `Enable` Property under Auth. We played with just setting the `DefaultAuthorizer: AWS_IAM` or `Auth.Authorizer: AWS_IAM` but forcing an Property to set what Authorizers are available made checking for Auth in other tools (like SAM CLI or CFN Lint) easier and more explicit.

Do you still have concerns in this approach?

jfuss reviewed (approved):
The looks good to me.

mgrandis commented:
All right, let's move on with this approach.

mgrandis commented on samtranslator/model/api/http_api_generator.py:
```diff
@@ -489,14 +489,21 @@ def _set_default_authorizer(self, open_api_editor, authorizers, default_authoriz
+            authorizers["AWS_IAM"] = ApiGatewayV2Authorizer(is_aws_iam_authorizer=True)
+
+        # If all the customer wants to do is enable the IAM authorizer the authorizers_config will be None.
+        if authorizers_config == None:
```
nit: Think it's better to use `is` instead of `==`

harrisonhjones replied:
ok, so `if authorizers_config is None` ? Why is `is` preferred over `==`?

mgrandis replied:
`==` is for equality, `is` is for identity.
It's a good practice when checking that something is not set, to use `is None`
You may actually want to use `if not authorizers_config:` here, as this will also check for values like empty list `[]` or empty object `{}` (which I think could be good checks here)
About is/==: https://stackoverflow.com/a/48120163

harrisonhjones replied:
Got it. Fairly easy change.

mgrandis commented on samtranslator/open_api/open_api.py:
```diff
@@ -415,7 +415,8 @@ def _set_method_authorizer(self, path, method_name, authorizer_name, authorizers
 
-            if authorizer_name != "NONE":
+            # Neither the NONE nor the AWS_IAM built-in authorizers support authorization scopes.
+            if authorizer_name != "NONE" and authorizer_name != "AWS_IAM":
```
nit: You can use `not in ["NONE", "AWS_IAM"]` here

harrisonhjones replied:
Fancy; seems like a reasonable change.

mgrandis commented on samtranslator/model/apigatewayv2.py:
```diff
@@ -174,6 +178,14 @@ def generate_openapi(self):
         authorizer_type = self._get_auth_type()
 
+        if authorizer_type == "AWS_IAM":
+            openapi = {
```
We should maybe return early here and in the `JWT` case.
So that we don't have to scan the rest of the function to see what we're doing with `openapi`.

harrisonhjones replied:
Yea. I didn't do it that way because we didn't do it that way for any of the other authorizers. I would prefer to make this change, if desired, in another smaller PR.

harrisonhjones replied:
I'm open to that. As a Go developer I prefer early returns but if elif is more python-style I'm open to it. Can we _not_ do it as part of this PR?

moelasmar replied:
I prefer moving the openapi object creation to separate private functions, but it is Ok for now.

mgrandis commented on samtranslator/model/api/http_api_generator.py:
```diff
@@ -489,14 +489,21 @@ def _set_default_authorizer(self, open_api_editor, authorizers, default_authoriz
+
         if not isinstance(authorizers_config, dict):
             raise InvalidResourceException(self.logical_id, "Authorizers must be a dictionary.")
 
```
Sorry not able to add a comment down below, but we should maybe add a condition in the `for` loop to ignore or fail on finding an `AWS_IAM` object in `authorizers_config.items()` so that we don't overwrite the one defined by `enable_iam_authorizer`

harrisonhjones replied:
I think that's fair. I was originally thinking we'd want to allow an overwrite but that was when this change was opt-out. Now that it is opt-in we shouldn't let users shoot themselves in the foot.

harrisonhjones replied:
Yea. if they already have an authorizer called "AWS_IAM" SAM doesn't clobber/overwrite it. The user's authorizer is used (though it likely won't help them).

AndersAsa commented:
Any timeline for this feature being merged?

mgrandis commented on tests/model/api/test_http_api_generator.py:
```diff
@@ -1,6 +1,7 @@
 from unittest import TestCase
 from mock import patch
 import pytest
+import json
```
nit: I think you can remove this as it's not used anymore

mgrandis commented:
@harrisonhjones It looks like there are only a few minor updates to do, do you have time to work on them?
Thanks!

wolfeidau commented:
@mgrandis looks like the author isn't able to address this feedback, can we merge this and follow up with another PR?

I would love to see this merged as it has been an issue for a while.

CrayTexas2018 commented:
Any timeline on when this PR will be merged?

harrisonhjones commented:
> @mgrandis looks like the author isn't able to address this feedback, can we merge this and follow up with another PR?
> 
> I would love to see this merged as it has been an issue for a while.

I can try to take a look at fixing this next week. Sorry for the delay getting back to y'all.

AndersAsa commented:
> > @mgrandis looks like the author isn't able to address this feedback, can we merge this and follow up with another PR?
> > I would love to see this merged as it has been an issue for a while.
> 
> I can try to take a look at fixing this next week. Sorry for the delay getting back to y'all.

Thank you!!!! This will be super helpful for me

ben-elsen commented:
> > @mgrandis looks like the author isn't able to address this feedback, can we merge this and follow up with another PR?
> > I would love to see this merged as it has been an issue for a while.
> 
> I can try to take a look at fixing this next week. Sorry for the delay getting back to y'all.

Hi @harrisonhjones, will you have some time in the near future to finish this? 
This would extremely helpful!

harrisonhjones commented:
I have addressed all PR feedback and added a few new tests. `make pr` passes

harrisonhjones commented on samtranslator/model/api/http_api_generator.py:
```diff
@@ -24,8 +24,8 @@
-AuthProperties = namedtuple("_AuthProperties", ["Authorizers", "DefaultAuthorizer"])
-AuthProperties.__new__.__defaults__ = (None, None)
+AuthProperties = namedtuple("_AuthProperties", ["Authorizers", "DefaultAuthorizer", "EnableIamAuthorizer"])
+AuthProperties.__new__.__defaults__ = (None, None, None)
```
Looking at this now: should the last `None` here actually be `False`?

jfuss replied:
`None` is falsy so they both work the same. It would only matter if `None` had some value to be considered different. For this, I don't think it strictly matters.

harrisonhjones replied:
Would it be better to use False instead of None largely to make it clear we expect a boolean value?

CrayTexas2018 commented:
> 

Thank you! Hopefully this gets merged! Really appreciate the effort.

harrisonhjones commented on samtranslator/model/api/http_api_generator.py:
```diff
@@ -494,14 +494,21 @@ def _set_default_authorizer(self, open_api_editor, authorizers, default_authoriz
         Returns all authorizers for an API as an ApiGatewayV2Authorizer object
         :param authorizers_config: authorizer configuration from the API Auth section
-        :param default_authorizer: name of the default authorizer
+        :param enable_iam_authorizer: if true add an "AWS_IAM" authorizer
```
s/true/True ?

jfuss replied:
Technically yes but personally don't think it strictly matters. The communication of the comment is the same.

harrisonhjones commented on tests/model/api/test_http_api_generator.py:
```diff
@@ -1,6 +1,7 @@
 from unittest import TestCase
 from unittest.mock import patch
 import pytest
+import json
```
reminder to self to actually remove this...

harrisonhjones reviewed (commented):
Three small comments; two for more experienced Python folks.

jfuss commented on samtranslator/model/eventsources/push.py:
```diff
@@ -1218,38 +1218,53 @@ def _add_auth_to_openapi_integration(self, api, editor):
-                ),
-            )
+        if method_authorizer == "AWS_IAM":
+            if not enable_iam_authorizer:
```
Why fail if IAM auth is not enabled?

harrisonhjones replied:
I guess we don't need to. In fact if they have a custom authorizer named "AWS_IAM" this code change might actually break them. I'll remove it.

jfuss commented on tests/translator/input/error_http_api_invalid_iam_auth.yaml:
```diff
@@ -0,0 +1,37 @@
+Resources:
```
I know you added integration tests (which thank you!) but could we add some of these input/output tests for valid cases? It is much easier to run these tests as quick validations of the translations.

harrisonhjones replied:
I'll look into that sure

jfuss reviewed (changes requested):
Just a couple small things.

harrisonhjones commented:
Addressed PR feedback; added five new tests, removed an unnecessary error test, and updated and existing multi-auth test to include the new iam authorizer.

CoshUS reviewed (commented):
Thank you for the contribution! One small question and everything else looks good to me.

CoshUS reviewed (approved):
LGTM! Thanks for addressing the comments!

harrisonhjones commented:
@CoshUS you are welcome :) What are the next steps to getting this merged? Anything I need to do?

harrisonhjones commented:
Rebased off of develop, fixed merge conflicts, updated new util function that was breaking tests, squashed into a single commit, ran `make pr` successfully, and force-pushed.

harrisonhjones commented on tests/translator/output/aws-cn/http_api_custom_iam_auth.json:
```diff
@@ -0,0 +1,137 @@
+          "components": {
+            "securitySchemes": {
+              "AWS_IAM": {
+                "type": "oauth2",
```
See the description in tests/translator/input/http_api_custom_iam_auth.yaml:

```
# This test-case tests what happens when an AWS_IAM authorizer is defined on an HttpApi but not enabled anywhere else.
# While the defined authorizer isn't really a true Iam authorizer (it's just a poorly-named OAuth authorizer) this shouldn't cause an error.
```

jfuss commented:
There are some things we need to handle internally (docs for example). I need to kick off that processes before merging.

claudiocidade commented:
great work guys this is fantastic. any estimates as to when it's merged/released? I have a bit of a life/death situation here that would greatly benefit from this one

brysontyrrell commented:
I was just talking to my TAM about this. 👀
